### DISCUSSION TITLE
Image component: ref, complete, loaded event

### DISCUSSION THREAD

**Discussion by endquote** [vercel/next.js#18386] [Ideas]
# Feature request

I'm looking for `ref` and `complete` properties on `next/Image`, as well as a `loaded` event.

## Is your feature request related to a problem? Please describe.

Congrats on launching Next 10! The new image component looks really neat. I was going to port my site to use it, but I'm not sure I can with the current version.

I have some components that watch for the image to be loaded, and then set some CSS to fade the container in. This requires setting a `ref` on `<img>`, accessing a `complete` property, and using the `onLoad` event. Is there a way to implement this with the new `next/Image` component?

```tsx
export const ContentCard = ({ image }:{ image: string }) => {
  const img = useRef<HTMLImageElement>(null!);

  const [loaded, setLoaded] = useState(false);

  useEffect(() => {
    if (img.current) {
      setLoaded(img.current.complete);
    }
  }, [img.current]);

  function imageLoaded(e: SyntheticEvent<HTMLImageElement, Event>) {
    if (img.current) {
      setLoaded(img.current.complete);
    }
  }

  return (
    <div className={loaded ? "loaded" : "loading"}>
      <img ref={img} src={image} onLoad={imageLoaded} />
    </div>
  );
};
```

## Describe the solution you'd like

Generally it seems like exposing more of the native `<img>` API in the `next/Image` component would be helpful.

**Comment by endquote** [comment]
I see. What is the proper way to type the ref? I thought the property wasn't there but it looks like I'm just typing it wrong.

![Screen Shot 2020-10-28 at 6 20 44 AM](https://user-images.githubusercontent.com/330010/97441309-d11f5100-18e5-11eb-96a3-716295bd1536.png)

**Reply by artecoop** [reply]
I'd love to know if you're come up with something about this

**Reply by lxsmnsyc** [reply]
The `<Image />` component cannot receive/forward a ref:

https://github.com/vercel/next.js/blob/canary/packages/next/client/image.tsx#L207-L220

**Reply by artecoop** [reply]
That's bummer. Hope they will add it soon

**Comment by g-elwell** [comment]
I gave up on this a while ago after having no joy using a ref, but it turns out that `onLoad` works great (as suggested above!)

A ref on the component will apply to it's outermost `div` wrapper, but `onLoad` appears to bubble down and attach itself to the image(s) it finds.

This triggers 2x per component as it also picks up the invisible placeholder that `next/image` uses, to avoid this just check that `srcset` exists as this will only be present on your 'real' image element.

eg:
```

const handleLoad = e => {
  if (e.target.srcset) {
    // your image is ready
  }
};

<NextImage {...props} onLoad={handleLoad} />
```

This works in conjunction with the IntersectionObserver and lazy loading within `next/image` since your images are only loaded when they enter the viewport.

I'm using `setState` and CSS to trigger an opacity change so my images fade in - works great!

**Reply by snacksncode** [reply]
Could you please explain in detail how did you get the onLoad to work? I still cannot really understand how to access onLoad event on their new Image component. I'm using Next v10.0.1

**Reply by dalalRohit** [reply]
> Could you please explain in detail how did you get the onLoad to work? I still cannot really understand how to access onLoad event on their new Image component. I'm using Next v10.0.1

You can use something like this:

```
const handleImgLoad=(e) => {
  //e.target contains all the DOM properties similar to ref.current
  e.target.classList.remove('hide-img');
  
  e.target.classList.add('show-img');
}
<Image className="hide-img" {...props} onLoad={handleImgLoad} />
```

Hope this helps.

**Reply by dalalRohit** [reply]
> I gave up on this a while ago after having no joy using a ref, but it turns out that `onLoad` works great (as suggested above!)
> 
> A ref on the component will apply to it's outermost `div` wrapper, but `onLoad` appears to bubble down and attach itself to the image(s) it finds.
> 
> This triggers 2x per component as it also picks up the invisible placeholder that `next/image` uses, to avoid this just check that `srcset` exists as this will only be present on your 'real' image element.
> 
> eg:
> 
> ```
> 
> const handleLoad = e => {
>   if (e.target.srcset) {
>     // your image is ready
>   }
> };
> 
> <NextImage {...props} onLoad={handleLoad} />
> ```
> 
> This works in conjunction with the IntersectionObserver and lazy loading within `next/image` since your images are only loaded when they enter the viewport.
> 
> I'm using `setState` and CSS to trigger an opacity change so my images fade in - works great!

Thank you so much for this. Was looking for this instead of a state updating solution. 🙌

**Comment by sigrundis** [comment]
Curious on the same problem, onLoad doesn't fire using the next/image component, and can't use ref.current.completed because the ref props don't pass down to the img element. Using version ^10.0.5.

I wrote my own progressive loading component but I want to use the optimized version of the image provided by the next/image component so I don't get the "Serve images in next-gen formats" warning at Google Pagespeed. Any suggestions?

**Comment by Quesstor** [comment]
If you are using the new `Image` component you can use this workaround
```tsx
const [loaded, setLoaded] = useState(false)
const ref = useRef<HTMLDivElement>(null)
useEffect(() => {
  if ((ref.current?.firstChild?.firstChild as HTMLImageElement | undefined)?.complete) {
    setLoaded(true)
  }
}, [])

return <div ref={ref}>
         <Image src={src} 
           onLoad={() => setLoaded(true)}/>
       </div>
```

**Comment by nuthinking** [comment]
After upgrading from `^12.0.7` to `^12.1.4` the `onLoad` event stopped working. We would like to have a sleek fade-in animation when the image is ready but we can't access reliably the `<img />` element (serverside rendering gets in the way of `id` based hacks). It's beyond my comprehension that `ref` is not supported and that `onLoadingComplete` doesn't pass any target. 🤯

How could you add animations to `Image` otherwise?
